Raoof/lok2 266 rbac editor has access to rooftops beyond assigned scope

raoofnaushad opened the pull request:
# What type of PR is this? (check all applicable)
- [ ] Refactor
- [x] Feature
- [x] Improvement
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
---
## Description
This PR implements comprehensive RBAC (Role-Based Access Control) restrictions for the **Viewer role**, completing the security framework for call management functionality. The implementation ensures that Viewers have read-only access to calls from their assigned rooftop while preventing any call creation, editing, or initiation actions.

### Key Changes:
- **Backend API Security**: Restricted call creation, initiation, and bulk upload endpoints to Editor+ roles
- **Frontend UI Protection**: Hidden restricted functionality and added access control guards
- **Enhanced Auth Context**: Added viewer-specific permission flags for granular access control
- **User Experience**: Clean, role-appropriate interfaces with helpful access denied messages

### Affected Areas:
- Call management endpoints (`/calls/*`)
- Demo call functionality
- Bulk upload system
- Frontend call components and pages
- Authentication and authorization system
---
## ENV / DB Changes
:warning: _If this PR introduces any **environment variable** or **database schema changes**, highlight them here._
- [ ] ENV changes
- [ ] DB changes
- Details: **No environment or database changes required** - this is purely an access control implementation using existing user roles and permissions.
---
## Screenshots / Recordings

### Admin Role (No Changes)
- Full access to all call functionality
- Can create, edit, initiate calls across all rooftops
- Can bulk upload calls

### Editor Role (No Changes) 
- Can create, edit, and initiate calls for assigned rooftop
- Can bulk upload calls for assigned rooftop
- Cannot access other rooftops

### Viewer Role (New Restrictions)
**Before**: Could create, edit, and initiate calls
**After**: Read-only access with clean UI that doesn't expose restricted functionality

#### Viewer - Calls Page
- Bulk Upload button is hidden
- Call action buttons are restricted
- Clean interface showing only viewable content

#### Viewer - Demo Call Section
- Shows AccessDenied component instead of initiation button
- Form fields remain viewable but non-interactive for restricted actions

#### Viewer - Bulk Upload Modal
- Shows AccessDenied message when modal is opened
- Prevents any upload functionality

---
## QA Instructions

### Prerequisites
1. Ensure you have test users with different roles:
   - **Admin User**: Full organizational access
   - **Editor User**: Assigned to a specific rooftop
   - **Viewer User**: Assigned to a specific rooftop
2. Test environment should have multiple rooftops with existing calls

### Test Scenarios

#### 1. **Admin Role Verification** (Should work as before)
1. Login as Admin user
2. Navigate to Calls page
3. ✅ Verify Bulk Upload button is visible and functional
4. ✅ Verify can create demo calls
5. ✅ Verify can initiate calls
6. ✅ Verify can switch between rooftops
7. ✅ Verify can access organization settings

#### 2. **Editor Role Verification** (Should work as before)
1. Login as Editor user
2. Navigate to Calls page
3. ✅ Verify Bulk Upload button is visible and functional
4. ✅ Verify can create demo calls for assigned rooftop
5. ✅ Verify can initiate calls for assigned rooftop
6. ✅ Verify cannot switch rooftops (locked to assigned rooftop)
7. ❌ Verify cannot access organization settings
8. ✅ Verify can access profile settings

#### 3. **Viewer Role Verification** (New Restrictions)
1. Login as Viewer user
2. Navigate to Calls page
3. ❌ **Verify Bulk Upload button is HIDDEN**
4. ✅ Verify can view existing calls from assigned rooftop
5. ✅ Verify can view call details and transcripts
6. Navigate to Demo Call section
7. ❌ **Verify demo call initiation shows AccessDenied message**
8. ❌ **Verify cannot create new calls**
9. ❌ **Verify cannot edit existing calls**
10. ❌ **Verify cannot initiate normal calls**

#### 4. **API Endpoint Security Testing**
1. Use browser dev tools or API client
2. Try to access restricted endpoints as Viewer:
   - `POST /api/v1/calls` (Create call) → Should return 403 Forbidden
   - `POST /api/v1/calls/{id}/initiate` (Initiate call) → Should return 403 Forbidden
   - `POST /api/v1/calls/demo/{id}/initiate` (Demo call) → Should return 403 Forbidden
   - `POST /api/v1/calls/bulk-upload` (Bulk upload) → Should return 403 Forbidden
   - `PUT /api/v1/calls/{id}` (Update call) → Should return 403 Forbidden

#### 5. **Cross-Role Navigation Testing**
1. Login as Admin, navigate around the app
2. Switch to Editor role, verify appropriate restrictions
3. Switch to Viewer role, verify all restrictions are applied
4. Verify no console errors or broken UI elements

#### 6. **Settings Page Access Testing**
1. Login as Viewer user
2. Navigate to Settings page
3. ✅ Verify can access Profile settings
4. ❌ Verify Organization settings section is HIDDEN
5. ❌ Verify cannot access billing, members, or integrations sections

#### 7. **Rooftop Scope Testing**
1. Login as Editor user assigned to Rooftop A
2. ✅ Verify can only see calls from Rooftop A
3. ❌ Verify cannot access calls from Rooftop B
4. Login as Viewer user assigned to Rooftop A
5. ✅ Verify can only see calls from Rooftop A (read-only)
6. ❌ Verify cannot access calls from Rooftop B

#### 8. **Error Handling and User Experience**
1. As Viewer, try to access restricted functionality
2. ✅ Verify AccessDenied components show helpful messages
3. ✅ Verify "Contact your administrator" guidance is shown
4. ✅ Verify no broken UI elements or console errors
5. ✅ Verify smooth navigation between allowed sections

### Expected Results Summary:
- **Admin**: Full access (no changes)
- **Editor**: Rooftop-scoped access for call management (no changes)
- **Viewer**: Read-only access with clean UI and helpful error messages

:white_check_mark: Tested on **Chrome/Firefox/Safari** on **macOS/Windows**
---
## Self-Review Checklist
- [x] I have tested the changes introduced in this PR
- [x] I have removed console.logs and debug logic before raising this PR
- [x] I have added/updated tests where possible
- [x] Code coverage target (≥ 70%) maintained
- [x] All linting errors have been resolved
- [x] Backend API endpoints properly secured with role-based dependencies
- [x] Frontend components have appropriate access control guards
- [x] User experience is clean and professional for all roles
---
## Review Process
**Two reviewers are required before merging:**
1. **Ramees (Functionality Review)**
   - Does the RBAC implementation meet security requirements?
   - Are all Viewer restrictions properly implemented?
   - Does the user experience remain professional and helpful?
   - Are there any edge cases or security gaps?
   - Does this impact existing Admin/Editor functionality?
2. **Suggested Reviewer (Code Review)**
   - Review backend endpoint security implementations
   - Check frontend component access control logic
   - Verify auth context enhancements are correct
   - Ensure no regressions in existing functionality
   - Review CallActionGuard component implementation
_Add the GitHub-suggested reviewer(s) for code review._
---
## Technical Implementation Details

### Backend Changes:
- **File**: `server/app/api/v1/endpoints/calls.py`
  - Updated dependencies from `get_current_user` to `get_editor_user_with_scope`
  - Added rooftop access verification for non-admin users
  - Secured: call creation, initiation, bulk upload, and update endpoints

### Frontend Changes:
- **File**: `client/src/contexts/AuthContext.tsx`
  - Added: `canInitiateCalls`, `canBulkUploadCalls`, `canCreateCalls`, `canEditCalls`
  - Role-based permission logic for granular access control

- **File**: `client/src/pages/Calls.tsx`
  - Conditional rendering of Bulk Upload button based on permissions

- **File**: `client/src/components/calls/DemoCallSection.tsx`
  - Access control for demo call initiation with AccessDenied fallback

- **File**: `client/src/components/calls/BulkUploadModal.tsx`
  - Modal-level access control with user-friendly error messages

- **File**: `client/src/components/auth/CallActionGuard.tsx` (New)
  - Reusable access control components for call-related actions
  - Convenience components: `CreateCallGuard`, `EditCallGuard`, etc.

### Security Considerations:
- All restrictions enforced at API level (backend security)
- Frontend UI provides clean user experience
- No sensitive functionality exposed to unauthorized users
- Proper error handling and user guidance

---
## :white_check_mark: Merge Requirements
- At least **2 approvals required**:
  - :heavy_check_mark: **Ramees** (Functionality Review)
  - :heavy_check_mark: **One suggested reviewer** (Code Quality Review)
- PR can be merged **only after**:
  - [x] ENV/DB changes (if any) are clearly documented ✅ No changes required
  - [x] QA instructions are verified ✅ Comprehensive test scenarios provided
  - [x] Self-review checklist is completed ✅ All items checked
  - [x] Security implications have been reviewed ✅ RBAC properly implemented
  - [x] User experience has been validated ✅ Clean, role-appropriate interfaces

raoofnaushad commented on server/app/api/v1/endpoints/calls.py:
```diff
@@ -926,12 +930,16 @@ async def update_call(
+    # First get the call to verify rooftop access
+    call = await CallService.get_call_by_id(call_id, organization.id, db)
+    if call and call.rooftop_id:
+        await verify_rooftop_access(call.rooftop_id, current_user, db)
```
Fixing it

raoofnaushad replied:
Fixed CallService Method Bug
Fixed AttributeError in update_call endpoint on line 940 - changed CallService.get_call_by_id() to CallService.get_call() to match the actual method name in the service class.
Before: call = await CallService.get_call_by_id(call_id, organization.id, db)
After: call = await CallService.get_call(call_id, organization.id, db)
This resolves a runtime error that would occur when updating calls.

raoofnaushad commented on server/app/dependencies.py:
```diff
@@ -153,6 +154,84 @@ async def role_checker(current_user: User = Depends(get_current_user)) -> User:
+        if current_user.rooftop_id:
+            return [current_user.rooftop_id]
+        return []
+
```
fixed it.

raoofnaushad commented:
@Muhammed-Saleeq  - all bugs fixed and updated based on our latest requiremnt. PRD also updated. Check this please.

FYI @zer0dark7 and @sinanshamsudheen

zer0dark7 reviewed (changes requested):
Editor [Access Denied ]- Go back button should redirect to org settings instead of  API platform.

raoofnaushad commented:
# What type of PR is this? (check all applicable)
- [ ] Refactor
- [x] Feature
- [x] Improvement
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
---
## Description
Improves UX for **editor users** in Settings page. Instead of showing unhelpful "Go Back" button on access-denied screens, editors are now redirected to **Organization Settings** where they have access.

**Problem**: Editor clicks "Members" → sees "Access Denied" → clicks "Go Back" → confused  
**Solution**: Editor clicks "Members" → sees "Go to Organization Settings" button → taken to accessible content

---
## ENV / DB Changes
- [ ] ENV changes
- [ ] DB changes
---
## Screenshots / Recordings

https://github.com/user-attachments/assets/f03a808f-1c93-4b41-b62a-a0ebb5c32ba8


---
## QA Instructions
**Test as Editor user:**
1. Login as editor (not admin)
2. Go to Settings → click "Members" or "Billing & Add-Ons"
3. ✅ Should see "Go to Organization Settings" button (not "Go Back")
4. Click button → should navigate to Org Settings tab
5. Try direct URLs: `/settings?section=billing` → should auto-redirect to org-settings

**Regression test:**
- Admin users: All sections still accessible
- Viewer users: Still see "Go Back" button

:white_check_mark: Tested on **Chrome** on **macOS**
---
## Self-Review Checklist
- [x] I have tested the changes introduced in this PR
- [x] I have removed console.logs and debug logic before raising this PR
- [ ] I have added/updated tests where possible
- [x] Code coverage target (≥ 70%) maintained
---
## Review Process
**Two reviewers are required before merging:**
1. **Ramees (Functionality Review)**
2. **Suggested Reviewer (Code Review)**

---
## :white_check_mark: Merge Requirements
- At least **2 approvals required**:
  - :heavy_check_mark: **Ramees** (Functionality Review)
  - :heavy_check_mark: **One suggested reviewer** (Code Quality Review)
- PR can be merged **only after**:
  - [x] ENV/DB changes (if any) are clearly documented
  - [x] QA instructions are verified
  - [x] Self-review checklist is completed

zer0dark7 reviewed (approved):
New changes look good.
Approving the PR 👍🏽


Thankyou Sir🫡

sinanshamsudheen reviewed (approved):
Bro this stuff is really good, I can't comprehend the amount of brainpower required for this. this single PR has taught me a lot regarding managing roles and adding safety constraints. great stuff!
(Im sure them merge conflicts will be crazy on this one)
